### ISSUE TITLE
What is the simplest way to convert trained Pytorch-weights of EfficientNet-Lite model back to Tensorflow? 

### ISSUE THREAD

**Issue by AlexeyAB** [rwightman/gen-efficientnet-pytorch#50]
@rwightman Hi,

What is the simplest way to convert trained Pytorch-weights of EfficientNet-Lite model back to Tensorflow? 
If I want to train EfficientNet-Lite model in Pytorch and then use it in TensorFlow (pb/tflite)?

I have done it by using https://github.com/onnx/onnx-tensorflow but it adds extra transpose layers and use multiple conv2d-layers instead of one GroupedConv2D, so Pytorch->ONNX->TensorFlow-model is 10x times slower than the native TensorFlow model: https://github.com/onnx/onnx-tensorflow/issues/782

**Comment by rwightman** [comment]
I've done a lot of TF -> PyTorch but been quite a while since I did the other way around. I've never used any utility, I transpose the weights myself and map the names. Mapping the names is a pain though because Tensorflow models often have annoying inconsistent naming due to the scoping and mix of scopes being manually set or some layers being left to auto-count the instances.  If you can somehow assign the weights in model forward order, instead of just passing unordered dict keys, you can rip through them sequentially and ignore the names. However I haven't figured out how to do that for TF1.

**Comment by AlexeyAB** [comment]
@rwightman 

I can successfully transfer VGG16 (feature extractor, except Avgpool/Dense layers) weights from Pytorch to TF, so both shows the same output results: https://colab.research.google.com/gist/AlexeyAB/36abb2f629dde486b0d49ea953e33572/weights_torch_to_tf_vgg16.ipynb
* [weights_torch_to_tf_vgg16_ipynb.zip](https://github.com/rwightman/gen-efficientnet-pytorch/files/5368822/weights_torch_to_tf_vgg16_ipynb.zip)
```
[[[0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  ...
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.18344408]]

 [[0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  ...
  [0.         0.         0.         ... 0.         0.18581158 0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]
  [0.         0.         0.         ... 0.         0.         0.        ]]
...
```

----

But when I do the same for EfficientNet-Lite-0 (I am trying to port only 3 layers `conv, dw_conv, conv`) then it shows different output results: https://colab.research.google.com/gist/AlexeyAB/b589f95ff03140cbc3f5af3389957bfc/weights_torch_to_tf_effnet_lite0.ipynb
* [weights_torch_to_tf_effnet_lite0_ipynb.zip](https://github.com/rwightman/gen-efficientnet-pytorch/files/5368826/weights_torch_to_tf_effnet_lite0_ipynb.zip)

The only difference is that there are no Batch-norm in VGG16, but it is in the EfficientNet-Lite0 (in Pytorch, but not in TF-model).

----

1. Are there any nuances that I should consider?

2. Why are there `scales, bias, mean, variance` in the Pytorch-model, while there are only `scales, bias` in TF-model?

3. Should I create TF-model in another way to have Batch-norm variables `scales, bias, mean, variance`  ?

4. Or should I fuse Batch-norm (scale, bias, mean, var) to the (scale, bias) only in such a way?
```
bias = bias - scale * rolling_mean / (np.sqrt(rolling_variance + .001))
scale = scale / (np.sqrt(rolling_variance + .001))
```

5. Can you share your script for TF -> Pytorch conversion?



6. Do I understand correctly that we should convert PT->TF weights in a such way:
* for conv `np.transpose(w, (2, 3, 1, 0))`: Pytorch `[n][c][ky][kx]`-> TF `[ky][kx][c][n]`
* for dw-conv `np.transpose(w, (2, 3, 0, 1))`: Pytorch `[n][c][ky][kx]`-> TF `[ky][kx][n][c]`

**Comment by rwightman** [comment]
Oh, you're working with TFlite? Those model defs are often optimized for inference, batch norm may be fused with neighbouring conv layer? 

I know the TF (not TF lite) definitions of those models definitely have gamma (aka weight or scale), beta (bias), and moving_mean (running_mean), moving_variance (running_var. I used the tensorflow training checkpoints for porting to PyTorch, not the .tflite model, otherwise it wouldn't have the BN layers useable for training.

**Comment by rwightman** [comment]
@AlexeyAB oh wait, nm, I read that wrong, you're just exporting tflite but calling the normal builder. Hmm, what BN layer are you using? Is it using the normal TF BN or the TPU specific one? Maybe that's the issue?

**Comment by AlexeyAB** [comment]
@rwightman 

>Hmm, what BN layer are you using? Is it using the normal TF BN or the TPU specific one? Maybe that's the issue?

May be yes, but how to use TF BN in a correct way?

I tried forcing to use `tf.layers.BatchNormalization`, but it didn't work, there are no still `mean, variance`.


I used:
```python
override_params = {}
override_params['batch_norm_momentum'] = 0.99
override_params['batch_norm_epsilon'] = 1e-3
override_params['batch_norm'] = tf.layers.BatchNormalization

features, endpoints, model = efficientnet_lite_builder.build_model(
        images, model_name, training=True, features_only=False, override_params=override_params, fine_tuning=False, model_dir="tfmodel"
)
```


Also I just changed `utils.TpuBatchNormalization` to `tf.layers.BatchNormalization` in several ways:
* https://github.com/tensorflow/tpu/blob/0ffe1274745806c411ed3dda7e84f692e00df8af/models/official/efficientnet/lite/efficientnet_lite_builder.py#L74
* https://github.com/tensorflow/tpu/blob/0ffe1274745806c411ed3dda7e84f692e00df8af/models/official/efficientnet/lite/efficientnet_lite_builder.py#L145

ALso I added line `override_params['batch_norm'] = tf.layers.BatchNormalization` between these two lines: https://github.com/tensorflow/tpu/blob/0ffe1274745806c411ed3dda7e84f692e00df8af/models/official/efficientnet/lite/efficientnet_lite_builder.py#L145-L146

**Comment by rwightman** [comment]
one other thing, are you just iterating over the trainable vars? moving mean/var won't be in there, you'd need the global_variables graphkey,  trainable ops are roughly == pytorch parameters, variables with gradients.

**Comment by AlexeyAB** [comment]
Oh thanks, this actually solved my problem! 
All that is needed:
1. `trainable_collection = tf.get_collection_ref(tf.GraphKeys.GLOBAL_VARIABLES)`
2. and
```
override_params = {}
override_params['batch_norm_momentum'] = 0.99
override_params['batch_norm_epsilon'] = 1e-3
override_params['batch_norm'] = tf.layers.BatchNormalization
```
3. and I added to fix and issue with `tf.FusedBatchNormV3` while converting to the TFLite model:
```
    converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,
                                       tf.lite.OpsSet.SELECT_TF_OPS]
```


https://colab.research.google.com/gist/AlexeyAB/92408ba118a156426fe6de3187d2bcd8/weights_torch_to_tf_effnet_lite0.ipynb
* [weights_torch_to_tf_effnet_lite0_ipynb.zip](https://github.com/rwightman/gen-efficientnet-pytorch/files/5369146/weights_torch_to_tf_effnet_lite0_ipynb.zip)

**Comment by AlexeyAB** [comment]
@rwightman 
But it doesn't seem to work correctly when I try to transfer weights for almost the entire model, not just the first three layers: https://colab.research.google.com/gist/AlexeyAB/605e4488933cf6c42406cb5648301ae6/weights_torch_to_tf_effnet_lite0.ipynb

**Comment by AlexeyAB** [comment]
Everything works well. Just training mode should be disabled `mode.eval()` for Pytorch and `training=False` for TF: https://colab.research.google.com/gist/AlexeyAB/cc05f2690c3707d5e0f66d1e749b5546/weights_torch_to_tf_effnet_lite0.ipynb#scrollTo=WFnIID6iBlsq

* [weights_torch_to_tf_effnet_lite0_ipynb.zip](https://github.com/rwightman/gen-efficientnet-pytorch/files/5372391/weights_torch_to_tf_effnet_lite0_ipynb.zip)

Thanks for help!
